Add docs for custom cell type references

sjspielman opened the pull request:
Closes #567 

This PR adds docs for DIY references! I mostly followed the outline I laid out for this here: https://github.com/AlexsLemonade/scpca-nf/issues/567#issuecomment-1934856023. I made a few small updates for things I noticed in the external instructions along the way, as well.

There's quite a bit here, so I suggest the first round of review be focused on overall scope and organization - is everything that needs to be said there and roughly presented in the right order?

Then in the next round(s) we can focus more on the wording and specifics.

allyhawkins commented on external-instructions.md:
```diff
@@ -402,7 +404,20 @@ We have provided an example project cell type metadata file for reference.
 
-When cell typing is turned on with `--perform_celltyping`, `scpca-nf` will skip annotation for any libraries whose cell type annotation results already exist in the `checkpoints` folder, as long as the cell type reference file is unchanged.
+When cell typing is turned on with `--perform_celltyping`, `scpca-nf` will skip annotation for any libraries whose cell type annotation results already exist in the `checkpoints` folder, as long as the cell type reference file name is unchanged.
+The cell type annotations in the `checkpoints` folder have the following structure:
```
```suggestion
The cell type annotations in the `checkpoints` folder will have the following structure:
```

allyhawkins commented on external-instructions.md:
````diff
@@ -402,7 +404,20 @@ We have provided an example project cell type metadata file for reference.
+```
+checkpoints
+└── celltype
+    └── library01
+    │   ├── library01_cellassign
+    │   └── library01_cellassign
+    └── library02
+        ├── library02_cellassign
+        └── library02_cellassign
````
```suggestion
    └── library01
    │   ├── library01_cellassign
    │   └── library01_singler
    └── library02
        ├── library02_cellassign
        └── library02_singler
```

allyhawkins commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,144 @@
+
+### Creating a custom `SingleR` reference dataset
+
+`SingleR` reference files should be saved as RDS files and named `<singler_reference_name>_model.rds`, where `<singler_reference_name>` is a string of your choosing.
```
I think before you talk about file types, you should mention what kind of data can be used as a reference for SingleR. Say that you can use bulk RNA-seq datasets or single cell datasets that already have annotations, noting that having annotations is the key piece of information. Also it would be good to mention resources where you can find other datasets (like scRNA-seq). I think we provide all of the celldex references, but you could mention that and then link to the docs that lists all included cell types in each celldex reference.

sjspielman replied:
Re-reading my text, I think I also need to expand more on the distinction between "reference" and "model" - they are not creating the reference, necessarily, but building a model from a reference they identify.

allyhawkins commented on custom-celltype-references.md:
````diff
@@ -0,0 +1,144 @@
+)
+```
+
+Once the model object is built, you can optionally add the field `reference_label` with information about the types of labels you specified.
+If you do not provide this information, the final `SingleCellExperiment` object produced by `scpca-nf` will have a value `NULL` for the field `metadata(sce)$singler_reference_label`, but no calculations will be affected.
+
+If you choose to add this label, use the following code:
+
+```
+singler_model$reference_label <- "<name describing my labels>"
````
I think I would mention both `reference_label` and `reference_name` as slots in the metadata. Also I would move the sentence about not setting it and it being NULL to after you show the code to do it.

allyhawkins replied:
Wait do we automatically grab the `reference_label` from the filename if it's not in the metadata of the object?

sjspielman replied:
So, this was a "cool" find actually: I'm pretty sure that it turns out we never use `reference_name`, but we sometimes use `reference_label` just to check of it's `label.ont` and handle ontology things accordingly. So I figured...let's just not tell them about these things? I also had the urge to remove `reference_name` from the object but controlled myself since one day maybe it will be useful

jashapiro replied:
Since we do use this, I think it is worth mentioning the ontology handling and what label will trigger that code.  I also want a plug somewhere encouraging people to use cell ontologies.

sjspielman replied:
> Wait do we automatically grab the reference_label from the filename if it's not in the metadata of the object?

We only do this in the workflow that builds the reference, so I don't think it's relevant here.

allyhawkins commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,144 @@
+
+
+
+`CellAssign` reference files should be saved as TSV files and named `<cellassign_reference_name>.tsv`, where `<cellassign_reference_name>` is a string of your choosing.
```
It might be worth a note both here and in the SingleR section about making sure the references are in the same organism for your sample, otherwise the gene IDs won't match up.

allyhawkins replied:
Again, I would also mention that a Cell Assign reference file comes from converting a list of marker genes into a binary matrix with all possible cell types as the columns and all possible genes as the rows. Some context before jumping into file type I think is helpful.

allyhawkins commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,144 @@
+1. `ensembl_id`, which contain Ensembl gene IDs.
+2. One column for each cell type label to use for annotation.
+All values in this column should be `1` or `0`, indicating whether that row's Ensembl gene id is a marker gene for that cell type (`1`) or not (`0`)
+3. Optionally, you may wish to a include a column called `"other"` with all values set to `0`.
```
I don't think this should be optional, otherwise CellAssign will assign it to one of the other cell types, so we want this column to exist.

allyhawkins commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,144 @@
+
+Several [reference datasets](./external-instructions.md#singler-references) and [marker gene lists](./external-instructions.md#cellassign-references) have been pre-compiled for use.
+
+This document provides instructions for creating and using your own reference data set for `SingleR` annotation or marker gene list for `CellAssign`, as well as setting up `scpca-nf` to run with these custom references.
```
I think in the overview I would include a numbered list of steps needed to run the workflow with a custom reference: 
1. Create the custom reference 
2. Create a cell type reference metadata file 
3. Run the workflow 

I would also reorganize the headers so those are the three main sections.

allyhawkins commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,144 @@
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**
+
+- [Creating and using custom cell type references](#creating-and-using-custom-cell-type-references)
+  - [Overview](#overview)
+    - [Creating a custom `SingleR` reference dataset](#creating-a-custom-singler-reference-dataset)
+      - [Training the model](#training-the-model)
+    - [Creating a custom `CellAssign` marker-gene list](#creating-a-custom-cellassign-marker-gene-list)
+    - [Creating your cell type metadata file](#creating-your-cell-type-metadata-file)
+    - [Running the workflow with custom references](#running-the-workflow-with-custom-references)
+    - [Special considerations for repeating cell type annotation](#special-considerations-for-repeating-cell-type-annotation)
```
I would move the doctoc to be below the first header and remove that from the TOC. 
Also maybe based on my suggested addition to the overview do this for the header organization: 
```suggestion
- [Overview](#overview)
- [Creating custom references for cell type annotation] 
    - [Creating a custom `SingleR` reference dataset](#creating-a-custom-singler-reference-dataset)
      - [Training the model](#training-the-model)
    - [Creating a custom `CellAssign` marker-gene list](#creating-a-custom-cellassign-marker-gene-list)
- [Creating your cell type metadata file](#creating-your-cell-type-metadata-file)
- [Running the workflow with custom references](#running-the-workflow-with-custom-references)
- [Special considerations for repeating cell type annotation](#special-considerations-for-repeating-cell-type-annotation)
```

allyhawkins commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,144 @@
+
+Use the following values for any projects where you'd like to use custom references:
+
+- `scpca_project_id`: A project ID.
+All libraries associated with this project ID will use the same cell type references.
+- `singler_ref_name`: The name of your `SingleR` model which should match the `<singler_reference_name>` string you used when creating your `SingleR` reference RDS file.
+If you do not wish to perform cell type annotation with `SingleR`, you can use `NA` in this column.
+- `singler_ref_file`: The file name of your `SingleR` model, as `<singler_reference_name>_model.rds`.
+If you do not wish to perform cell type annotation with `SingleR`, you can use `NA` in this column.
+- `cellassign_ref_name`: The name of your `CellAssign` model which should match the `<cellassign_ref_name>` string you used when creating your `CellAssign` reference TSV file.
+If you do not wish to perform cell type annotation with `CellAssign`, you can use `NA` in this column.
+- `cellassign_ref_file`: The file name of your `CellAssign` model, as `<cellassign_reference_name>.tsv`.
+If you do not wish to perform cell type annotation with `CellAssign`, you can use `NA` in this column.
```
I would turn this into a table, similar to how we talk about the metadata files in the other instructions.

allyhawkins commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,144 @@
+
+When performing cell type annotation, `scpca-nf` requires a TSV metadata file with information about cell type references, [as described in these instructions](./external-instructions.md#preparing-the-project-cell-type-metadata-file).
+
+Use the following values for any projects where you'd like to use custom references:
```
```suggestion
This metadata should include the following columns. Each row should indicate a `scpca_project_id` that you would like to perform cell typing with a custom reference. 
```
Maybe not this exact wording, but something like this.

allyhawkins commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,144 @@
+There are several parameter values you will have to set either in your [configuration file](./external-instructions.md#configuration-files) or at the command line when running `scpca-nf`:
+
+
+- `params.singler_reference_label`: A name describing the label used during `SingleR` model training.
+If you do not have a particular name for your labels, you can simply provide `""`, _but you MUST provide a non-default value for this parameter_ unless you are not performing cell type annotation with `SingleR`.
+- `params.singler_models_dir`: The path/uri to the directory that contains your `<singler_reference_name>_model.rds`.
+This value should _not include_ the file name itself.
+For example, if the path to your file is `/path/to/reference/<singler_reference_name>_model.rds`, only provide the value `/path/to/reference`.
+You can skip this parameter if you are not performing cell type annotation with `SingleR`.
+- `params.cellassign_reference_dir`: The path/uri to the directory that contains your `<cellassign_reference_name>.tsv`.
+This value should _not include_ the file name itself.
+For example, if the path to your file is `/path/to/reference/<cellassign_reference_name>.tsv`, only provide the value `/path/to/reference`.
+You can skip this parameter if you are not performing cell type annotation with `CellAssign`.
+- `params.project_celltype_metafile`: The path/uri to your cell type metadata file, including the the file name.
+
```
We might want to include these in the config template we provide and encourage users to change them in the config file.

allyhawkins reviewed (commented):
Thanks for tackling this! I left a bunch of organizational comments throughout to get us started.

jashapiro commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,144 @@
+To train your model, use the following code, where
+
+- `ref` is your `SummarizedExperiment` or `SingleCellExperiment` reference dataset object with columns as samples or cells and rows as genes.
+- `labels` is a named vector providing labels for the reference dataset, where names correspond to the column names in your reference object.
```
I don't think this has to be named (and the name description is pretty confusing!)
```suggestion
- `labels` is a vector providing cell type labels for the sample or cells in the reference dataset
```

jashapiro reviewed (commented):
I just looked very briefly, which started as a search for "onto" to see if you said anything about ontologies. I want us to say something encouraging people to use ontologies for their cell line labels. It may fall on deaf ears much of the time, but I want the little nudges where we can.

jashapiro commented on examples/user_template.config:
```diff
@@ -79,3 +81,14 @@ profiles{
 }
+
+
+/*
+Parameters to set when performing cell type annotation with custom references
+------------------------------
+The values presented here are examples, and should be replaced with
+the correct paths to any indicated files or other values that need to be set.
+*/
```
I would expect these parameters to be be above the computational requirements section, keeping all parameters together.

sjspielman commented:
Thanks everyone for the initial round of feedback! I've implemented reviews and expanded instructions where recommended, did a bit of reorg, I removed some spots with instructions that didn't need to be there, and I added some params to the example config. Ready for round two!

jashapiro reviewed (commented):
I was looking because I wanted to see what precommit ci had done here, and I had just a couple of quick comments, before further review.

allyhawkins commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,195 @@
+Using this specific string will trigger the `scpca-nf` workflow to additionally expect a dataframe with two columns, `ontology_id` and `ontology_cell_names`, to be present in the `SingleR` model object in the field `celltype_ontology_df`.
+Therefore, if you supply the string `"label.ont"`, you must also add this field.
+The `ontology_id` column should contain ontology id values corresponding to labels in your `labels` vector, and the `ontology_cell_names` column should contain the human-readable cell type names associated with each ontology ID (e.g., `"B cell"`).
+The [`oncoProc` package](https://bioconductor.org/packages/release/bioc/html/ontoProc.html) from `Bioconductor` may be useful for compiling this information.
```
```suggestion
The [`ontoProc` package](https://bioconductor.org/packages/release/bioc/html/ontoProc.html) from `Bioconductor` may be useful for compiling this information.
```

allyhawkins commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,195 @@
+
+#### Special considerations when using ontology labels
+
+If you chose to use ontology labels, you may wish to set the `reference_label` field to the string `"label.ont"` as shown below.
```
Should we tell them where to get cell ontology labels from?

allyhawkins replied:
Also it may be nice to include a brief sentence about why we encourage ontology labels.

sjspielman replied:
FYI I added this higher up in the docs where I discuss where you can find ontology ids

allyhawkins commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,195 @@
+
+The `CellAssign` reference file should be created by converting a list of marker genes for a set of cell type labels into a binary matrix with values of `0` and `1`.
+This matrix should have all possible cell types as the columns and all possible genes, represented as Ensembl gene ids, as the rows.
+Values of `1` indicate that the given gene is a marker gene for the give cell type, and values of `0` indicate that the gene is not a marker gene for the cell type.
```
```suggestion
Values of `1` indicate that the given gene is a marker gene for the given cell type, and values of `0` indicate that the gene is not a marker gene for the cell type.
```

allyhawkins commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,195 @@
+The `CellAssign` reference file should be created by converting a list of marker genes for a set of cell type labels into a binary matrix with values of `0` and `1`.
+This matrix should have all possible cell types as the columns and all possible genes, represented as Ensembl gene ids, as the rows.
+Values of `1` indicate that the given gene is a marker gene for the give cell type, and values of `0` indicate that the gene is not a marker gene for the cell type.
+When compiling this information, it is important to use marker genes from the same organism as the sample you are analyzing to ensure matching gene ids.
```
Also they should be ensembl gene ids.

allyhawkins commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,195 @@
+
+When performing cell type annotation, `scpca-nf` requires a TSV metadata file with information about cell type references, [as described in these instructions](./external-instructions.md#preparing-the-project-cell-type-metadata-file).
+
+All libraries a given `scpca_project_id` will use the same reference dataset for each of `SingleR` and `CellAssign`, respectively.
```
```suggestion
All libraries with the same `scpca_project_id` will use the same reference dataset for each of `SingleR` and `CellAssign`, respectively.
```

allyhawkins commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,195 @@
+
+When running the workflow, be sure to use the `--perform_celltyping` flag to perform cell type annotation.
+
+You can run the workflow as (showing all parameters specified at the command line, for example):
```
```suggestion
You can run the workflow using the below code. 
The example below shows all parameters specified at the command line, but if specifying them in the configuration file you would be able to remove those lines. 
```
Just being a little bit more explicit here.

allyhawkins commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,195 @@
+
+
+As described [in these instructions](./external-instructions.md#repeating-cell-type-annotation), cell type annotation is, by default, not repeated if results already exist.
+`scpca-nf` determines if results already exist based on the _file name_ of the provided reference file.
```
I think you also want to mention that its either the file name not changing or the existence of intermediate checkpoint files. At first I looked at this and was like that's not correct, but then I remembered we check both.

allyhawkins reviewed (commented):
I like this organization a lot! I just had a few more small comments.

sjspielman commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,195 @@
+## Create custom cell type references
+
+
+### Using a custom `SingleR` reference dataset
```
I'm glad you pointed this out, since I felt torn on the language. They aren't exactly creating a _reference_ so much as creating a _model_ based on an existing reference. 

As I write this, I wonder - is the answer actually just, `Creating a `SingleR` model from a custom reference dataset`?

allyhawkins replied:
> As I write this, I wonder - is the answer actually just, Creating a SingleR model from a custom reference dataset?

I like this!

sjspielman commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,195 @@
+Rows should be named with Ensembl gene ids.
+    a. Some resources you can use to find a suitable reference dataset are the [`celldex` package](https://bioconductor.org/packages/release/data/experiment/vignettes/celldex/inst/doc/userguide.html) or the [`scRNA-seq` package](https://bioconductor.org/packages/release/data/experiment/html/scRNAseq.html), both from Bioconductor.
+Note that the [`SingleR` reference datasets which `scpca-nf` has pre-compiled for use](./external-instructions.md#singler-references) are a selected subset of references from the `celldex` package.
+2. Use [`SingleR::traingSingleR()`](https://rdrr.io/github/LTLA/SingleR/man/trainSingleR.html) to train your model based on your chosen reference dataset.
```
Amazingly either of us can spell 😂

allyhawkins commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,218 @@
+
+
+As described [in these instructions](./external-instructions.md#repeating-cell-type-annotation), cell type annotation is, by default, not repeated if results already exist.
+`scpca-nf` determines if results already exist based on two criteria: i) the _file name_ of the provided reference file, and ii) the existence of [intermediate checkpoint files](./external-instructions.md#repeating-cell-type-annotation) from previous runs of cell typing.
```
```suggestion
`scpca-nf` determines if results already exist based on two criteria: i) the _file name_ of the provided reference file has not changed, and ii) [intermediate checkpoint files](./external-instructions.md#repeating-cell-type-annotation) from previous runs of cell typing already exist.
```

allyhawkins reviewed (approved):
Just one minor comment, but I don't need to see this again.

jashapiro commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,218 @@
+
+### Creating a custom `CellAssign` reference from a marker-gene list
+
+The `CellAssign` reference file should be created by converting a list of marker genes for a set of cell type labels into a binary matrix with values of `0` and `1`.
```
```suggestion
The `CellAssign` reference file is created by converting a list of marker genes for a set of cell type labels into a binary matrix with values of `0` and `1`.
```

jashapiro commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,218 @@
+### Creating a custom `CellAssign` reference from a marker-gene list
+
+The `CellAssign` reference file should be created by converting a list of marker genes for a set of cell type labels into a binary matrix with values of `0` and `1`.
+This matrix should have all possible cell types as the columns and all possible genes, represented as Ensembl gene ids, as the rows.
```
```suggestion
This matrix should have all possible cell types as the columns and all possible genes, represented as Ensembl gene IDs, as the rows.
```

jashapiro commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,218 @@
+The `CellAssign` reference file should be created by converting a list of marker genes for a set of cell type labels into a binary matrix with values of `0` and `1`.
+This matrix should have all possible cell types as the columns and all possible genes, represented as Ensembl gene ids, as the rows.
+Values of `1` indicate that the given gene is a marker gene for the given cell type, and values of `0` indicate that the gene is not a marker gene for the cell type.
+When compiling this information, you should use marker genes (represented as Ensembl gene ids) from the same organism as the sample you are analyzing to ensure matching Ensembl gene ids.
```
```suggestion
When compiling this information, be sure to use Ensembl gene IDs from the same organism as the sample you are analyzing!
```

jashapiro commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,218 @@
+
+This TSV file should have the following columns and values:
+
+1. `ensembl_id`, which contains Ensembl gene ids.
```
```suggestion
1. `ensembl_id`, which contains Ensembl gene IDs.
```

jashapiro commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,218 @@
+[As described in the `SingleR` book](https://bioconductor.org/books/release/SingleRBook/), `SingleR` uses these reference expression values and associated labels to identify cells in your dataset with similar expression patterns.
+Either bulk gene expression (microarray or RNA-seq) or single-cell datasets with annotations can be used as references.
+To the best of your ability, you should identify and use a reference dataset that contains cell types that you expect will also be present in your data.
+The reference dataset should also be the same organism as the sample you are analyzing to ensure matching gene ids.
```
```suggestion
The reference dataset should also be the same organism as the sample you are analyzing to ensure matching gene IDs.
```

jashapiro commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,218 @@
+
+1. Identify an appropriate reference dataset to use.
+This dataset should be either a [`SummarizedExperiment`](https://rdrr.io/bioc/SummarizedExperiment/man/SummarizedExperiment-class.html) or [`SingleCellExperiment`](https://rdrr.io/bioc/SingleCellExperiment/man/SingleCellExperiment.html) object, where each column represents a sample or cell, and each row represents a gene.
+Rows should be named with Ensembl gene ids.
```
```suggestion
Rows should be named with Ensembl gene IDs.
```

jashapiro commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,218 @@
+
+| column_id             | contents |
+| --------------------- | -------- |
+| `scpca_project_id`    | Project id matching values in the run metadata file |
```
```suggestion
| `scpca_project_id`    | Project ID matching values in the run metadata file |
```

jashapiro reviewed (commented):
I left a few more comments, with the main one being a slight reorg with an extra subheading in the SingleR section. It just felt strange to have subheadings for some but not all steps.

sjspielman commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,218 @@
+This dataset should be either a [`SummarizedExperiment`](https://rdrr.io/bioc/SummarizedExperiment/man/SummarizedExperiment-class.html) or [`SingleCellExperiment`](https://rdrr.io/bioc/SingleCellExperiment/man/SingleCellExperiment.html) object, where each column represents a sample or cell, and each row represents a gene.
+Rows should be named with Ensembl gene ids.
+Some resources you can use to find a suitable reference dataset are the Bioconductor packages [`celldex`](https://bioconductor.org/packages/release/data/experiment/vignettes/celldex/inst/doc/userguide.html) and [`scRNA-seq`](https://bioconductor.org/packages/release/data/experiment/html/scRNAseq.html), and the [`Azimuth`](https://azimuth.hubmapconsortium.org/references/) database.
+Note that the [`SingleR` reference datasets which `scpca-nf` has pre-compiled for use](./external-instructions.md#singler-references) are a selected subset of references from the `celldex` package.
```
I'm not sure that showing how to work with `SummarizedExperiment` (or `SCE`s for that matter) is exactly the right scope here, but I think it would make sense to maybe also include links in this broken-out section, like

https://bioconductor.org/books/3.18/OSCA.intro/the-singlecellexperiment-class.html

and maybe this for `SE`? to be honest I don't think this one is communicating what we want to communicate though, but not sure about a different/better reference.... Any ideas? https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html

sjspielman replied:
I went ahead and added here to start fyi - https://github.com/AlexsLemonade/scpca-nf/pull/692/commits/9a5102d33bbeb90546b8bcd3354e680d787d1fb7

jashapiro replied:
Yes, I think links make the most sense here. But I would focus on conversion rather than the structure of the object. 

So for SummarizedExperiment I might link to https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html#constructing-a-summarizedexperiment

It isn't perfect, because it talks a lot about intervals, which we don't need, but it gets the idea that if you have a matrix, it is pretty easy to make the object, especially the second example (which should be first).

For SCE, maybe links to https://satijalab.org/seurat/reference/as.singlecellexperiment and https://bioconductor.org/packages/3.19/bioc/html/zellkonverter.html

sjspielman replied:
eh? https://github.com/AlexsLemonade/scpca-nf/blob/817c259d4c25dfd97bdfbf820397658090df6a4d/custom-celltype-references.md#identify-an-appropriate-reference-dataset

jashapiro commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,227 @@
+To the best of your ability, you should identify and use a reference dataset that contains cell types that you expect will also be present in your data.
+The reference dataset should be from the same organism as the sample you are analyzing.
+
+Your reference dataset should be formatted as either a [`SummarizedExperiment`](https://rdrr.io/bioc/SummarizedExperiment/man/SummarizedExperiment-class.html) or [`SingleCellExperiment`](https://rdrr.io/bioc/SingleCellExperiment/man/SingleCellExperiment.html) object, where each column represents a sample or cell, and each row represents a gene.
+Rows should be named with Ensembl gene IDs.
+For more information on creating these types of objects, please see the following resources:
+
+- [Constructing a `SummarizedExperiment` object](https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html#constructing-a-summarizedexperiment)
+- An introduction to [the `SingleCellExperiment` class](https://bioconductor.org/books/3.18/OSCA.intro/the-singlecellexperiment-class.html#storing-primary-experimental-data)
+- [Converting `Seurat` objects](https://satijalab.org/seurat/reference/as.singlecellexperiment) or [`AnnData` objects](https://bioconductor.org/packages/3.19/bioc/html/zellkonverter.html) to `SingleCellExperiment` objects
+
+Some resources you may find helpful for obtaining a suitable reference dataset are the Bioconductor packages [`celldex`](https://bioconductor.org/packages/release/data/experiment/vignettes/celldex/inst/doc/userguide.html) and [`scRNA-seq`](https://bioconductor.org/packages/release/data/experiment/html/scRNAseq.html), and the [`Azimuth`](https://azimuth.hubmapconsortium.org/references/) database.
+Note that the [`SingleR` reference datasets which `scpca-nf` has pre-compiled for use](./external-instructions.md#singler-references) are a selected subset of references from the `celldex` package.
```
I reversed the order here to keep the section about what a reference dataset is with where to find them, followed by the conversion section, along with just a bit of wording changes. (should -> must)

```suggestion
Some resources you may find helpful for obtaining a suitable reference dataset are the Bioconductor packages [`celldex`](https://bioconductor.org/packages/release/data/experiment/vignettes/celldex/inst/doc/userguide.html) and [`scRNA-seq`](https://bioconductor.org/packages/release/data/experiment/html/scRNAseq.html), and the [`Azimuth`](https://azimuth.hubmapconsortium.org/references/) database.
Note that the [`SingleR` reference datasets which `scpca-nf` has pre-compiled for use](./external-instructions.md#singler-references) are a selected subset of references from the `celldex` package.

For use with `SingleR`, the reference dataset must be formatted as either a [`SummarizedExperiment`](https://rdrr.io/bioc/SummarizedExperiment/man/SummarizedExperiment-class.html) or [`SingleCellExperiment`](https://rdrr.io/bioc/SingleCellExperiment/man/SingleCellExperiment.html) object, where each column represents a sample or cell, and each row represents a gene.
Rows should be named with Ensembl gene IDs.

For more information on creating these types of objects, please see the following resources:

- [Constructing a `SummarizedExperiment` object](https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html#constructing-a-summarizedexperiment)
- An introduction to [the `SingleCellExperiment` class](https://bioconductor.org/books/3.18/OSCA.intro/the-singlecellexperiment-class.html#storing-primary-experimental-data)
- [Converting `Seurat` objects](https://satijalab.org/seurat/reference/as.singlecellexperiment) or [`AnnData` objects](https://bioconductor.org/packages/3.18/bioc/html/zellkonverter.html) to `SingleCellExperiment` objects
```

jashapiro commented on custom-celltype-references.md:
```diff
@@ -0,0 +1,227 @@
+3. [Add additional fields](#add-additional-fields-to-the-model-object) to the established model object
+4. [Export the trained model](#export-the-model) to a file named `<singler_reference name>_model.rds`
+
+#### Identify an appropriate reference dataset
```
Maybe?
```suggestion
#### Selecting and formatting a reference dataset
```

jashapiro reviewed (approved):
Just a small additional update to change order, along with a suggestion about the header to include the formatting scope. (I didn't update the link, because you might not choose this title.)

Other than that, all looks good.
